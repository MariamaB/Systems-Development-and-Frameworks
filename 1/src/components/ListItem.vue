<template>
  <li>
    <div class="itemContainer">
      <div class="messageSection" style="display:inline-block;">
        <template v-if="onEdit">
          <input v-model="newTodo.message" />
                  <select v-model="newTodo.assignedTo">
                    <option>assign to ...</option>
                  <option 
                     v-for="(user, index) in users"
                    :key="index"
                    :value="user.id" 
                     :selected="user.id === newTodo.assignedTo"
                     >
                     {{ user.email }}
                     </option>
                </select>
                
        </template>
        <template v-else><span style="font-weight: bold;">{{todo.message}}</span>  - assignt to: {{assignedToUser}}</template>
      </div>

      <div class="mb5 space">
        <span class="buttonContainer" v-if="!onEdit">
          <a @click="setEditMode(true)">
            <img title="Edit" class="icon" src="../assets/edit_circ.png" />
          </a>
          <span class="space" />
          <a @click="deleteListItem(todo.id)">
            <img title="Delete" class="icon" src="../assets/delete_circ.png" />
          </a>
        </span>
        <span class="buttonContainer" v-if="onEdit">
          <a @click="saveChanges()">
            <img title="Save" class="icon" src="../assets/save_circ.png" />
          </a>
          <span class="space" />
          <a @click="setEditMode(false)">
            <img title="Cancel" class="icon" src="../assets/cancel_circ.png" />
          </a>
        </span>
      </div>
      <div v-if="checkLoggedInUser">
        <a v-if="this.todo.status" ><img title="Done" class="icon" src="../assets/done.png" /></a>
        <a v-else @click="setTodoDone"><img title="Done_Pending" class="icon" src="../assets/done_pending.png" /></a>
      </div>
    </div>
  </li>
</template>

<script>
import { clone } from 'lodash'
import gql from 'graphql-tag'

export default {
  name: "list-item",
  props: {
    users: {
      type: Array,
      required: true
    },   
    todo: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      onEdit: false,
      newTodo: {
        message: '',
        assignedTo: null,
        status: false
      }
    };
  },
  methods: {
    setEditMode(value) {
      if(value)
        this.newTodo = clone(this.todo);
        
      this.onEdit = value;
    },
    saveChanges() {
      this.setEditMode(false);
      this.$emit("todoChanged", this.newTodo);
    },
    changeMessage(event) {
      this.newMsg = event.target.value;
    },
    deleteListItem(id) {
      this.$emit("deleteListItem", id);
    },
    async setTodoDone(){
      await this.$apollo.mutate({
      mutation: gql`mutation ($id: String!, $status: Boolean!) {
        changeTodoStatus(id: $id, status: $status) {
          status
        }
      }`,
      // Parameters
      variables: {
        id: this.todo.id,
         status: true,
      },
    }).then((data) => {
                    if (data.data.changeTodoStatus != null || data.data.changeTodoStatus!= undefined) {
                    this.todo.status = data.data.changeTodoStatus.status;
                    }
                    
                }).catch((error) => {
                    return error;
                })
      },
  },
  computed: {
     assignedToUser() {
       /* eslint-disable */
        console.log(this.todo.assignedTo)
        console.log(this.newTodo.assignedTo)
       let user = this.users.find((u) => u.id === this.todo.assignedTo);
       return user === undefined ? "none" : user.email
     },

     checkLoggedInUser(){
      if(this.todo.assignedTo != null){
        let user = this.users.find(u => u.id === this.todo.assignedTo);
        return user.loggedIn ? true : false;    
      }else{
        return false;
      }   
  }
  }
};
</script>

<style>
.itemContainer {
  border-bottom-style: inset;
  margin-bottom: 20px;
  border-bottom-color: #42b983;
  display: flex;
}

.buttonContainer {
  display: flex;
  cursor: pointer;
}

.mb5 {
  margin-bottom: 5px;
}

.messageSection {
  display: flex;
  align-self: center;
  justify-self: center;
  margin-left: 15px;
  flex: 2;
}

.icon {
  width: 30px;
  height: 30px;
}

.space {
  margin-right: 20px;
}

a {
  display: flex;
  align-content: center;
}

input {
  color: #42b983;
}
</style>