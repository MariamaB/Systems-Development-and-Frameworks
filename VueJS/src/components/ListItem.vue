<template>
  <li>
    <div class="itemContainer">
      <span class="messageSection">
        <span v-if="onEdit">
          <input @input="changeMessage" />
        </span>
        <span v-else>{{message}}</span>
      </span>

      <div class="mb5">
        <span class="buttonContainer" v-if="!this.onEdit">
          <a @click="setEditMode(true)">
            <img class="icon" src="../assets/edit_circ.png" />
          </a>
          <span class="space" />
          <a @click="deleteListItem(id)">
            <img class="icon" src="../assets/delete_circ.png" />
          </a>
        </span>
        <span class="buttonContainer" v-if="this.onEdit">
          <a @click="saveChanges()">
            <img class="icon" src="../assets/save_circ.png" />
          </a>
          <span class="space" />
          <a @click="setEditMode(false)">
            <img class="icon" src="../assets/cancel_circ.png" />
          </a>
        </span>
      </div>
    </div>
  </li>
</template>

<script>
export default {
  name: "list-item",
  props: {
    id: String,
    message: String
  },
  data() {
    return {
      onEdit: false,
      newMsg: ""
    };
  },
  methods: {
    setEditMode(value) {
      this.onEdit = value;
    },
    saveChanges() {
      this.setEditMode(false);

      if (this.newMsg.trim().length === 0) alert("Enter sth");
      else this.$emit("messageChanged", this.newMsg);
    },
    changeMessage(event) {
      this.newMsg = event.target.value;
    },
    deleteListItem(id) {
      this.$emit("deleteListItem", id);
    }
  }
};
</script>

<style scoped>
.editSection {
  flex-direction: row;
  flex: 4;
}

.itemContainer {
  border-bottom-style: inset;
  margin-bottom: 20px;
  border-bottom-color: #42b983;
  display: flex;
  width: 30%;
}

.buttonContainer {
  display: flex;
}

.mb5 {
  margin-bottom: 5px;
}

.messageSection {
  display: flex;
  align-self: center;
  justify-self: center;
  margin-left: 15px;
  flex: 2;
}

.icon {
  width: 30px;
  height: 30px;
}

.space {
  margin-right: 20px;
}

a {
  display: flex;
  align-content: center;
}

input {
  color: #42b983;
}
</style>