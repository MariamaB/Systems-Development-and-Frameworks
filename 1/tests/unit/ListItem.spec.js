import { shallowMount } from '@vue/test-utils';
import ListItem from '@/components/ListItem.vue';
import { mount } from '@vue/test-utils';

describe('ListItem.vue', () => {
	it('render props.message when passed', () => {
		const message = 'new todo';
		const wrapper = shallowMount(ListItem, {
			propsData: { message }
		});
		expect(wrapper.text()).toMatch(message);
	});
});

describe('ListItem', () => {
	describe('given a `todo`', () => {
		const wrapper = mount(ListItem, {
			name: 'item',
			propsData: {
				id: 1,
				message: 'Foo'
			}
		});

		
		it('renders `item.message`', () => {
			expect(wrapper.vm.message).toBe('Foo');
		});
		it("doesn't show input field", () => {
			expect(wrapper.vm.onEdit).toBeFalsy();
		});
		it('has the expected html structure', () => {
			expect(wrapper.element).toMatchSnapshot();
		});

		describe('Click on `Edit` button', () => {
			it('shows input field', () => {
				const aArray = wrapper.findAll('a');
				const Edit = aArray.at(0);
				Edit.trigger('click');
				expect(wrapper.vm.onEdit).toBeTruthy();
			});

			describe('edit text and submit', () => {
				it('$emit `Save` with edited todo', () => {
					wrapper.setProps({ id: 1, message: 'Moi' }); //since OnEdit is true we can overwrite item
					const aArray = wrapper.findAll('a');
					const Save = aArray.at(1); // because of  v-if condition Save button is now second element
					Save.trigger('click');
					expect(wrapper.vm.onEdit).toBeFalsy();
					expect(wrapper.vm.message).toBe('Moi');
				});
				it('has the expected html structure', () => {
					expect(wrapper.element).toMatchSnapshot(); // 2 Snapshot are produced to examine the state before and after edit, to see state while editting "Save click" should be commented and except onEdit must be Truthy
				});
			});

			describe('testing `Delete` button', () => {
				it('click on button emits delete event', () => {
					wrapper.setProps({ id: 1, message: 'Test'});
					expect(wrapper.vm.message).toBe('Test');
					const aArray = wrapper.findAll('a');
					const Delete = aArray.at(1);
					Delete.trigger('click');
					expect(wrapper.find(wrapper.vm.message).exists()).toBe(false);
				
				});
			});


	

			//Note in Snapshot just 2 icons can be seen according to the statement v-if="!this.onEdit"
		});
	});
});
