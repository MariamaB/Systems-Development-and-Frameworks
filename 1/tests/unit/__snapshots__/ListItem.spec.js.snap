// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem given a \`todo\` Click on \`Edit\` button edit text and submit has the expected html structure 1`] = `
<li>
  <div
    class="itemContainer"
  >
    <span
      class="messageSection"
    >
      <span>
        Moi
      </span>
    </span>
     
    <div
      class="mb5"
    >
      <span
        class="buttonContainer"
      >
        <a>
          <img
            class="icon"
            src="../assets/edit_circ.png"
            title="Edit"
          />
        </a>
         
        <span
          class="space"
        />
         
        <a>
          <img
            class="icon"
            src="../assets/delete_circ.png"
            title="Delete"
          />
        </a>
      </span>
       
      <!---->
    </div>
  </div>
</li>
`;

exports[`ListItem given a \`todo\` has the expected html structure 1`] = `
<li>
  <div
    class="itemContainer"
  >
    <span
      class="messageSection"
    >
      <span>
        Foo
      </span>
    </span>
     
    <div
      class="mb5"
    >
      <span
        class="buttonContainer"
      >
        <a>
          <img
            class="icon"
            src="../assets/edit_circ.png"
            title="Edit"
          />
        </a>
         
        <span
          class="space"
        />
         
        <a>
          <img
            class="icon"
            src="../assets/delete_circ.png"
            title="Delete"
          />
        </a>
      </span>
       
      <!---->
    </div>
  </div>
</li>
`;
